net/src/ip/prefix.rs

```diff
@@ -0,0 +1,419 @@
+use std::net::IpAddr;
+use std::str::FromStr;
+
+/// An `IpAddr` with a mask describing a network in CIDR notation.
+///
+/// Note that unlike [`IpNet`] from the `ipnet` crate, this type ensures that only network bits
+/// are set in the address.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub enum IpPrefix {
+    /// An IPv4 prefix
+    V4(Ipv4Prefix),
+    /// An IPv6 prefix
+    V6(Ipv6Prefix),
+}
+impl From<IpAddr> for IpPrefix {
+    fn from(value: IpAddr) -> Self {
+        match value {
+            IpAddr::V4(addr) => IpPrefix::V4(Ipv4Prefix::from(addr)),
+            IpAddr::V6(addr) => IpPrefix::V6(Ipv6Prefix::from(addr)),
+        }
+    }
+}
+
+/// A checked type describing possible lengths of ip prefixes
+pub enum IpPrefixLen {
+    /// A valid ipv4 prefix length
+    V4(Ipv4PrefixLen),
+    /// A valid ipv6 prefix length
+    V6(Ipv6PrefixLen),
+}
+
+/// Errors which may occur when evaluating the length of ip prefixes.
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidIpPrefixLength {
+    /// Prefix length too long for ipv4
+    #[error(transparent)]
+    InvalidIpv4PrefixLength(#[from] InvalidIpv4PrefixLength),
+    /// Prefix length too long for ipv6
+    #[error(transparent)]
+    InvalidIpv6PrefixLength(#[from] InvalidIpv6PrefixLength),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidIpNetwork {
+    /// An invalid Ipv4 network
+    #[error(transparent)]
+    InvalidIpv4Network(#[from] InvalidIpv4Network),
+    /// An invalid Ipv6 network
+    #[error(transparent)]
+    InvalidIpv6Network(#[from] InvalidIpv6Network),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum IpPrefixParseError {
+    /// A failed attempt to parse an ipv4 prefix
+    #[error(transparent)]
+    V4(#[from] Ipv4PrefixParseError),
+    /// A failed attempt to parse an ipv6 prefix
+    #[error(transparent)]
+    V6(#[from] Ipv6PrefixParseError),
+    /// Failed to parse input as an ipv4 or ipv6 prefix
+    #[error("failed to parse input '{0}' as ipv4 or ipv6 prefix")]
+    ParseFailure(String),
+}
+
+impl Display for IpPrefix {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            IpPrefix::V4(p) => write!(f, "{p}"),
+            IpPrefix::V6(p) => write!(f, "{p}"),
+        }
+    }
+}
+
+impl From<Ipv4PrefixLen> for IpPrefixLen {
+    fn from(value: Ipv4PrefixLen) -> Self {
+        IpPrefixLen::V4(value)
+    }
+}
+
+impl From<Ipv6PrefixLen> for IpPrefixLen {
+    fn from(value: Ipv6PrefixLen) -> Self {
+        IpPrefixLen::V6(value)
+    }
+}
+
+impl From<Ipv4Prefix> for IpPrefix {
+    fn from(value: Ipv4Prefix) -> Self {
+        IpPrefix::V4(value)
+    }
+}
+
+impl From<Ipv6Prefix> for IpPrefix {
+    fn from(value: Ipv6Prefix) -> Self {
+        IpPrefix::V6(value)
+    }
+}
+
+impl IpPrefix {
```

Comment:
This looks redundant with `routing::prefix::Prefix`? Do we need to keep both versions?

Reply:
Ok we chatted about it and I get it now - the plan is to move the `routing`'s version into the `net` crate instead.